### PR TITLE
8325880: Require minimum Open XL C/C++ version 17.1.1

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change that updates the minimum supported version of IBM
Open XL C/C++.  SAP dropped support for older versions in JDK 22, only
supporting the version specified in this change.

I need someone from the aix-ppc porters to test and review the change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325880](https://bugs.openjdk.org/browse/JDK-8325880): Require minimum Open XL C/C++ version 17.1.1 (**Enhancement** - P4)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17857/head:pull/17857` \
`$ git checkout pull/17857`

Update a local copy of the PR: \
`$ git checkout pull/17857` \
`$ git pull https://git.openjdk.org/jdk.git pull/17857/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17857`

View PR using the GUI difftool: \
`$ git pr show -t 17857`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17857.diff">https://git.openjdk.org/jdk/pull/17857.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17857#issuecomment-1944951272)

### CODE REVIEW

**Review by TheRealMDoerr — APPROVED** [review]
LGTM and works for us.
@JoKern65: We should probably remove the old build pipeline after this is integrated.

File: make/autoconf/toolchain.m4 [hunk 1]
```diff
@@ -53,7 +53,7 @@ TOOLCHAIN_DESCRIPTION_xlc="IBM XL C/C++"
 TOOLCHAIN_MINIMUM_VERSION_clang="3.5"
 TOOLCHAIN_MINIMUM_VERSION_gcc="6.0"
 TOOLCHAIN_MINIMUM_VERSION_microsoft="19.28.0.0" # VS2019 16.8, aka MSVC 14.28
-TOOLCHAIN_MINIMUM_VERSION_xlc="16.1.0.0011"
+TOOLCHAIN_MINIMUM_VERSION_xlc="17.1.1.4"
```
Comment (JoKern65) [inline make/autoconf/toolchain.m4:56]: We do not build AIX with the xlc toolchain any more but the clang one. So this line only stops a build if someone is trying to build with xlc 16 against toolchain xlc.
I have to agree to @TheRealMDoerr, that the correct change would be to remove the xlc toolchain in jdk23 at all.

**Review by TheShermanTanker — APPROVED** [review]

**Comment by magicus** [comment]
What does this mean? That you are not using xlc at all? Or is it clang but still with an xlc frontend, so all xlc flags etc need to stay?

**Review by magicus — APPROVED** [review]
Change look fine from a build perspective.

**Comment by JoKern65** [comment]
> What does this mean? That you are not using xlc at all? Or is it clang but still with an xlc frontend, so all xlc flags etc need to stay?


The `xlc` toolchain is for the compiler versions up to 16 (xlclang++); the `clang` toolchain is for the compiler versions 17 (ibm-clang++_r) and higher. For the 17 Compiler the frontend is `clang`-ish and we are using the `clang` flags instead of the `xlc` flags.
`toolchain.m4` decides on the basis of the found compiler which toolchain to use as default.
```
# On AIX the default toolchain depends on the installed (found) compiler
  #   xlclang++     -> xlc toolchain
  #   ibm-clang++_r -> clang toolchain
  # The compiler is searched on the PATH and TOOLCHAIN_PATH
  # xlclang++ has precedence over ibm-clang++_r if both are installed
```
So, if we set the minimum compiler level for AIX to 17, we can remove the xlc toolchain at all. 
We cannot remove every reference to xlc, because at least some headers we still use the xlc version (globalDefinitions_xlc.hpp)

**Comment by magicus** [comment]
@JoKern65 Thanks for the explanation! It would be nice to clear out the xlc stuff. Let's open a separate issue for that once this is integrated.

I also believe that the selection of `globalDefinitions_xlc.hpp` is done by #ifdefs in Hotspot source code, and has no relation to the build systems concept of toolchains.

**Comment by TheShermanTanker** [comment]
> @JoKern65 Thanks for the explanation! It would be nice to clear out the xlc stuff. Let's open a separate issue for that once this is integrated.
> 
> I also believe that the selection of `globalDefinitions_xlc.hpp` is done by #ifdefs in Hotspot source code, and has no relation to the build systems concept of toolchains.

There is one spot in the build system where clang is forcefully labelled as xlc, in the place where the HotSpot toolchain type is set
https://github.com/openjdk/jdk/blob/a0e5e16afbd19f6396f0af2cba954225a357eca8/make/autoconf/toolchain.m4#L1015

**Comment by magicus** [comment]
I see. I believe I wrote that piece of code, but I'd clearly forgotten that. 😕  Thanks! :)

**Comment by JoKern65** [comment]
> I see. I believe I wrote that piece of code, but I'd clearly forgotten that. 😕 Thanks! :)

No, this was added by me, because this was the root point to resolve to globalDefinitions_xlc.hpp

**Comment by TheShermanTanker** [comment]
> > I see. I believe I wrote that piece of code, but I'd clearly forgotten that. 😕 Thanks! :)
> 
> No, this was added by me, because this was the root point to still resolve to globalDefinitions_xlc.hpp even with toolchain clang

Shame that we can't fully swap to clang in some areas for AIX, but oh well

**Comment by JoKern65** [comment]
> > > I see. I believe I wrote that piece of code, but I'd clearly forgotten that. 😕 Thanks! :)
> > 
> > 
> > No, this was added by me, because this was the root point to still resolve to globalDefinitions_xlc.hpp even with toolchain clang
> 
> Shame that we can't fully swap to clang in some areas for AIX, but oh well

Although the compiler is now clang, the headers are still the old IBM ones and globalDefinitions_xlc.hpp is not consumable by other clang implementations. So if we change this we still have to differentiate between AIX clang and  other clangs.

**Comment by magicus** [comment]
Sounds like it really should be `globalDefinitions_aix.hpp` then...

**Comment by kimbarrett** [comment]
There was one bit of discussion about this tool chain upgrade over in the PR for enabling C++17:
https://github.com/openjdk/jdk/pull/14988#issuecomment-1741985638
Has that been sufficiently answered?  Good to go here?  @aixtools

**Comment by kimbarrett** [comment]
> > What does this mean? That you are not using xlc at all? Or is it clang but still with an xlc frontend, so all xlc flags etc need to stay?
> 
> The `xlc` toolchain is for the compiler versions up to 16 (xlclang++); the `clang` toolchain is for the compiler versions 17 (ibm-clang++_r) and higher. For the 17 Compiler the frontend is `clang`-ish and we are using the `clang` flags instead of the `xlc` flags. `toolchain.m4` decides on the basis of the found compiler which toolchain to use as default.
> 
> ```
> # On AIX the default toolchain depends on the installed (found) compiler
>   #   xlclang++     -> xlc toolchain
>   #   ibm-clang++_r -> clang toolchain
>   # The compiler is searched on the PATH and TOOLCHAIN_PATH
>   # xlclang++ has precedence over ibm-clang++_r if both are installed
> ```
> 
> So, if we set the minimum compiler level for AIX to 17, we can remove the xlc toolchain at all. We cannot remove every reference to xlc, because at least some headers we still use the xlc version (globalDefinitions_xlc.hpp)

This suggests there might be more that needs to be done here than simply
updating TOOLCHAIN_MINIMUM_VERSION_xlc.  I spent some time looking at the
relevant code, but keep getting lost in the distinction between xlc and clang.
Does updating that variable as proposed even work at all?

I'm going to need some help from you aix-ppc maintainer folks.

**Comment by TheRealMDoerr** [comment]
Your current version still serves the purpose of disallowing usage of xlc 16 and below. Users of the old compiler will get to know which compiler to use at minimum.
Users of any Open XL compilers are not affected because "TOOLCHAIN_MINIMUM_VERSION_xlc" is not used for that AFAICS. But that's ok IMHO because the minimum clang check will be in place and the supported compilers are documented (https://wiki.openjdk.org/display/Build/Supported+Build+Platforms).
So, I think this PR could get integrated and the old build pipeline get removed in a separate PR. Do you agree @JoKern65?

**Comment by JoKern65** [comment]
> > > What does this mean? That you are not using xlc at all? Or is it clang but still with an xlc frontend, so all xlc flags etc need to stay?
> > 
> > 
> > The `xlc` toolchain is for the compiler versions up to 16 (xlclang++); the `clang` toolchain is for the compiler versions 17 (ibm-clang++_r) and higher. For the 17 Compiler the frontend is `clang`-ish and we are using the `clang` flags instead of the `xlc` flags. `toolchain.m4` decides on the basis of the found compiler which toolchain to use as default.
> > ```
> > # On AIX the default toolchain depends on the installed (found) compiler
> >   #   xlclang++     -> xlc toolchain
> >   #   ibm-clang++_r -> clang toolchain
> >   # The compiler is searched on the PATH and TOOLCHAIN_PATH
> >   # xlclang++ has precedence over ibm-clang++_r if both are installed
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So, if we set the minimum compiler level for AIX to 17, we can remove the xlc toolchain at all. We cannot remove every reference to xlc, because at least some headers we still use the xlc version (globalDefinitions_xlc.hpp)
> 
> This suggests there might be more that needs to be done here than simply updating TOOLCHAIN_MINIMUM_VERSION_xlc. I spent some time looking at the relevant code, but keep getting lost in the distinction between xlc and clang. Does updating that variable as proposed even work at all?
> 
> I'm going to need some help from you aix-ppc maintainer folks.

As I already mentioned, This PR is just the start shot to remove the support for the old xlc compilers below version 17. This means removing the xlc toolchain support in a follow up PR by our team. This is feasible, because the open xl compilers starting with version 17 are using the clang toolchain. So, if this PR is through we feel empowered to remove the xlc toolchain.

**Comment by kimbarrett** [comment]
> Your current version still serves the purpose of disallowing usage of xlc 16 and below. Users of the old compiler will get to know which compiler to use at minimum. Users of any Open XL compilers are not affected because "TOOLCHAIN_MINIMUM_VERSION_xlc" is not used for that AFAICS. But that's ok IMHO because the minimum clang check will be in place and the supported compilers are documented (https://wiki.openjdk.org/display/Build/Supported+Build+Platforms). So, I think this PR could get integrated and the old build pipeline get removed in a separate PR. Do you agree @JoKern65?

The non-use of `TOOLCHAIN_MINIMUM_VERSION_xlc` when using the version 17 compilers is one of the things
that worried me.  (So I did correctly figure out it wouldn't be used.)  I think this means one could inadvertently use
17.1.0 rather than the desired requirement of 17.1.1.

I think raising the minimum clang version to 13 (as proposed in https://github.com/openjdk/jdk/pull/17862) doesn't
catch that either, since 17.1.0 is based on clang 13.

But if you folks are okay with this as an interim step, then I will go ahead.

**Comment by magicus** [comment]
It sounds like a good first step, yes. Let's push this and then clean out the xlc toolchain.

**Comment by kimbarrett** [comment]
No new objections, so going ahead with this.

**Pull request closed without merging**